simulation/navigator_2Dsim/sim2d.py

```diff
@@ -0,0 +1,252 @@
+#!/usr/bin/env python
+
+from __future__ import division
+import numpy as np
+import rospy
+import matplotlib.pyplot as plt
+import sys
+from sys import argv
+import tf.transformations as trns
+import ast
+import mil_tools
+
+from geometry_msgs.msg import WrenchStamped, PoseStamped, PointStamped, PoseWithCovariance, TwistWithCovariance
+from roboteq_msgs.msg import Command
+from nav_msgs.msg import Odometry, OccupancyGrid
+from rawgps_common import gps
+
+import rosbag
+from std_msgs.msg import Int32, String
+
+# Ros params that are now handled in launch.
+# Uncomment if you need to run this without the launch file.
+# rospy.set_param('inertia', ([1, 1, 2]))
+# rospy.set_param('drag', ([.5, 1, 1]))
+# rospy.set_param('mass', 544)
+
+# Twist seems off, also no covariance matrices? Pose looks fine tho.
+
+
+class Navsim():
+
+    def __init__(self, t=0, pose0=np.array([0, 0, 0]), twist0=np.array([0, 0, 0]), wind=None):
+
+        rospy.init_node('2Dsim')
+        self.des_force = np.array([0, 0, 0])
+        self.gui = rospy.get_param("gui")
+        self.world_frame_id = "/enu"
+        self.body_frame_id = "/base_link"
+        self.state = None
+        self.get_ref = None
+        self.last_odom = None
+        self.last_absodom = None
+        initial_lla = ast.literal_eval(rospy.get_param("start_pos"))
+        # print(intial_lla)
+        # self.last_ecef = gps.ecef_from_latlongheight(*np.radians(initial_lla))
+        self.last_enu = None
+        self.state_sub_max_prd = rospy.Duration(1 / 100)
+        self.last_state_sub_time = rospy.Time.now()
+
+        self.subscriber = rospy.Subscriber(
+            "/wrench/cmd", WrenchStamped, self.wrench_cb)
+        self.odompublisher = rospy.Publisher(
+            "/odom", Odometry, queue_size=1)
+        self.poses = []
+        self.wrenches = []
+        self.times = []
+        self.twists = []
+        # X, Y, and angle in world frame
+        self.pose = np.float64(initial_lla)
+        # X, Y, and rate of angle change in boat frame (Surge, Sway, yaw rate)
+        self.twist = np.float64(twist0)
+        # kg, kg, kg*m^2
+        self.inertia = np.array(ast.literal_eval(
+            rospy.get_param('inertia')), dtype=np.float64)
+        # Boat is longer than shorter, hence .5, 1, 1.
+        self.drag = np.array(ast.literal_eval(
+            rospy.get_param('drag')), dtype=np.float64)
+        # Time in seconds
+        self.t = t
+        # Wind is a function of time that returns a wrench
+        if wind:
+            self.wind = wind
+        else:
+            self.wind = lambda t: np.zeros(3, dtype=np.float64)
+
+        rospy.Timer(rospy.Duration(1 / 100), self.publish_odom)
```

Comment:
Wouldn't it make more sense to just have one timer that updates state and publishes odom instead of two?